testharness.js

```diff
@@ -527,11 +527,14 @@ policies and contribution forms [3].
             tests.promise_tests = Promise.resolve();
         }
         tests.promise_tests = tests.promise_tests.then(function() {
+            var earlyDonePromise = new Promise(function(resolve) {
+                test.add_cleanup(resolve);
+            });
             var promise = test.step(func, test, test);
             test.step(function() {
                 assert_not_equals(promise, undefined);
             });
-            return Promise.resolve(promise)
+            return Promise.race([promise, earlyDonePromise])
```

Comment:
I think it's a bit weird to call done() again after earlyDonePromise resolves. Possibly cleaner (and just as correct) would be to get rid of the "return" on this line, and add a "return earlyDonePromise" (maybe renamed to just "donePromise") after it? That way it's clear that the next test always is started when done is called, and the block here is just to make sure done is called when the passed in promise resolves.